### PR TITLE
[DOTNET] Create Utility class for Regular Expression

### PR DESCRIPTION
**Pull request by GiulianoDolceTalianoSantoro**
Solves [ISSUE#1427](https://github.com/microsoft/Recognizers-Text/issues/1427)

### Description
This PR will centralize all Regular expressions related methods in a class called `RegExpUtility` in .NET. The idea here is to increase parity to match how it is already implemented in Java, JavaScript and Python. 

### Changes made

- We added four more methods from `RegexExtensions.cs`:
   - `IsExactMatch( )`
   - `MatchExact( )`
   - `MatchEnd( )`
   - `MatchBegin( )`
 
- We changed `MatchEvaluator( )` to `ReplaceMatchGroup( )` because that name was more descriptive

- We moved the `ConditionalMatch` class from `Microsoft.Recognizers.Text.DateTime` to `Microsoft.Recognizers.Text.Utilities`, so the methods we moved before do not break the existing references.
 
- We imported `Microsoft.Recognizers.Text.Utilities` everywhere the code had errors because of the classes and methods migration.

### Testing
Find below a screenshot of the running tests

image

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs [hunk 1]
```diff
@@ -359,11 +359,11 @@ protected ParseResult FracLikeNumberParse(ExtractResult extResult)
 
                 var smallValue = char.IsDigit(numerator[0]) ?
                     GetDigitalValue(numerator, 1) :
-                    GetIntValue(GetMatches(numerator));
+                    GetIntValue(Utilities.RegExpUtility.GetMatches(this.TextNumberRegex, numerator)); // ACAACA
```
Comment (VictorGrycuk) [inline .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: Is the comment `// ACAACA` used for something?

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs [hunk 2]
```diff
@@ -696,21 +696,21 @@ private bool SkipNonDecimalSeparator(char ch, int distance)
...
-            while (successMatch.Success)
-            {
-                var matchStr = successMatch.Groups[0].Value;
-                matchStrs.Add(matchStr);
-                successMatch = successMatch.NextMatch();
-            }
-
-            return matchStrs;
-        }
+        // private List<string> GetMatches(string input)
```
Comment (VictorGrycuk) [inline .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: This whole method was commented, shouldn't it be removed instead?

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs [hunk 3]
```diff
@@ -696,21 +696,21 @@ private bool SkipNonDecimalSeparator(char ch, int distance)
...
+        // // Store all match str.
+        //    while (successMatch.Success)
+        //    {
+        //        var matchStr = successMatch.Groups[0].Value;
+        //        matchStrs.Add(matchStr);
+        //        successMatch = successMatch.NextMatch();
+        //    }
+
+        // return matchStrs;
+        // }
```
Comment (santgr11) [inline .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: Should this part be commented or deleted?

Comment (Maihav) [inline .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: Why is this method commented?

**Review by Maihav — CHANGES_REQUESTED** [review]

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs (hunk 3 above)
Comment (ceciliaavila) [inline .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: Please remove commented code.

**Comment by Maihav** [comment]
Remember to fix the documentation typo; it is `Utilities` not `Utilites`

**Review by ceciliaavila — CHANGES_REQUESTED** [review]
Just some minor feedback and a question. Why are you deleting .py files in here?

**Review by Konigstiger — APPROVED** [review]
LGTM, with the already proposed modifications of 
- innecesary comment.
- remove code commented method.

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs (hunk 1 above)
Comment (GiulianoDolceTalianoSantoro) [reply .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: It was only to remember to return to that line when looking for "ACAACA" in the code, but I forgot to delete it. I already solved it, thanks for the feedback.

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs (hunk 2 above)
Comment (GiulianoDolceTalianoSantoro) [reply .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: It should, I just forgot about it, but I removed it already, thank you!

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs (hunk 3 above)
Comment (GiulianoDolceTalianoSantoro) [reply .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: I didn't suppose to let that commented, but I forgot. I already removed it, thanks for the feedback!

Comment (GiulianoDolceTalianoSantoro) [reply .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: For some reason I actually see those files in my branch, I don't really know why they appear as deleted in the first commit (matcher.py & abstract_matcher.py). I'll let that for now as it is, thank you for pointing that and for the feedback!

image

Comment (GiulianoDolceTalianoSantoro) [reply .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: The problem was solved by deleting both Matcher and matcher folders (matcher was not supposed to be created, but for some reason it created itself as a Matcher clone, with the files that were supposedly deleted, an issue in Microsoft's Recognizers-Text repository) and adding Matcher again with the deleted Python files (as they were locally in my Matcher folder). We're going to make a PR to fix this in MS repository. Thanks for the feedback and for making us realize that.

MS & SW repository: 

image

feature/create-regex-utility-class branch (PR #304)

image

**Pull request closed without merging**